### PR TITLE
tracing: separate custom tag interface from trace driver

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

Commit Message: tracing: separate custom tag interface from trace driver
Additional Description: This is part of https://github.com/envoyproxy/envoy/pull/18955.
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by wbpcode** [comment]
Does this PR just move code from trace driver to separate source file? 👀

**Comment by Shikugawa** [comment]
> Does this PR just move code from trace driver to separate source file? 👀
Yes. This is needed to support custom tag in access logging.

**Comment by Shikugawa** [comment]
@envoyproxy/envoy-maintainers Could anyone take a look?

File: source/common/tracing/custom_tag_impl.cc [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  if (!metadata) {
+    return absl::nullopt;
+  }
+
+  const ProtobufWkt::Value& value = Envoy::Config::Metadata::metadataValue(metadata, metadata_key_);
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::kBoolValue:
+    return value.bool_value() ? "true" : "false";
+  case ProtobufWkt::Value::kNumberValue:
+    return absl::StrCat("", value.number_value());
```
Comment (rojkov) [inline source/common/tracing/custom_tag_impl.cc:68]: nit:
```suggestion
    return absl::StrCat(value.number_value());
```

**Review by rojkov — COMMENTED** [review]
Looks very good! Added one question to clarify.

File: source/common/tracing/custom_tag_impl.cc [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+void MetadataCustomTag::applySpan(Span& span, const CustomTagContext& ctx) const {
+  const envoy::config::core::v3::Metadata* meta = metadata(ctx);
+  const auto meta_str = metadataToString(meta);
+
+  if (!meta_str.has_value()) {
+    if (!default_value_.empty()) {
+      span.setTag(tag(), default_value_);
+    }
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (Shikugawa) [reply source/common/tracing/custom_tag_impl.cc:52]: Maybe it is unexpected behavior that the empty default value is inserted into the span.

[the remark this answers is not in the document]
Comment (rojkov) [reply source/common/tracing/custom_tag_impl.cc:52]: Ok, it's safer to preserve the current behavior for now as I don't have much experience with tracing.

**Review by rojkov — APPROVED** [review]
Thank you! LGTM

**Review by lizan — APPROVED** [review]

**Pull request merged**